### DISCUSSION TITLE
Making use of post-processing without blowing out the textures

### DISCUSSION THREAD

**Discussion by sortofsleepy** [google/filament#5757] [Q&A] [answered]
👋 

I have what I hope a very simple question - basically I would like to make use of the built-in DoF effect that's part of the engine; the goal is basically to animate the scene and have it go from blurry to clear.

However, the issue I run into is that turning on post-processing appears to run another effect(s) at the same time; the end result is that the textures used in the scene appear to be blown out and way over-exposed.

After searching for quite awhile, I came across another issue that suggested the Color Grading settings may play a part in why things look over exposed; I've tried to tweak things a little bit but haven't quite figured out the correct settings just yet. 

Can anyone point me in the right direction as to how to stop the textures from looking over-exposed? I could just roll my own bluring material and a render target or two, but since I can achieve what I'm looking for with DoF and it's built in already, I figured I may as try to save myself the trouble. 

Here are the current color grading settings. This helps a tiny bit but things still look noticeably over-exposed compared to the original image. (project is on Android in case that isn't already clear from the code snippit) 

```kotlin
   var colorGrading = ColorGrading.Builder()
           .toneMapper(ToneMapper.Linear())
           .saturation(1.1f)
           .exposure(0.1f)
           .build(engine)

    // DoF settings  - just testing values but trying to animate in from SUPER blurry to clear
   options.enabled = true
   options.cocScale = 2.8f / 0.8f
   options.nativeResolution = true;
   options.maxForegroundCOC = 140
   options.maxApertureDiameter = 50000f
   options.foregroundRingCount = 5

   camera.focusDistance = 90.0f;
```

Thank you!

**Comment by romainguy** [comment, marked as answer]
Did you turn off all post processing before? Doing so turned off "gamma correction", an important step. What you are seeing has nothing to do with exposure, but with linear vs color space encoded rendering. Sharing code snippets of your setup and screenshots would help.

**Reply by sortofsleepy** [reply]
> Did you turn off all post processing before? Doing so turned off "gamma correction", an important step. What you are seeing has nothing to do with exposure, but with linear vs color space encoded rendering. Sharing code snippets of your setup and screenshots would help.

Ah, yes I probably should have called that out specifically but post-processing was originally off in the project. And yes I realize that what I am seeing technically has nothing to do with exposure, I just didn't know how best to describe it, thank you for mentioning the correct terminology, I'll start searching again with that in mind.

 The project is an internal one and has quite a few moving pieces so it's difficult to share snippets but the scene I'm working on is relatively simple and essentially is along the lines of 
 
 ```kotlin
    
    //.... renderer object obtained from internal framework , assume "renderer" variable exists
    
    val camera: Camera = engine.createCamera(engine.entityManager.create())
    val clearOptions = Renderer.ClearOptions()
    var scene: Scene = engine.createScene()
    var view: View = engine.createView()
    var dofOptions = View.DepthOfFieldOptions()



    clearOptions.clear = true
    renderer.clearOptions = clearOptions


    /**
     * Loads a material that just displays a texture and with the following settings
    shadingModel : unlit,
    vertexDomain: object,
    flipUV : false,
    blending : transparent,
    depthCulling : true,
    depthWrite : true
     */
    private val material: Material = loadMaterial()

    var renderable = EntityManager.get().create()
    /// .... build out geometry with material and add to scene

    camera.setExposure(
        16.0f
        1.0f / 250.0f
        100.0f
    )


    view.camera = camera
    view.scene = scene
    view.isPostProcessingEnabled = true

    //// setting up DoF to try out animation ////
    dofOptions.enabled = true
    dofOptions.cocScale = 2.8f / 0.8f
    dofOptions.maxForegroundCOC = 140
    dofOptions.maxApertureDiameter = 50000f

    camera.focusDistance = 90.0f;

    var colorGrading = ColorGrading.Builder()
        .toneMapper(ToneMapper.Linear())
        .saturation(1.1f)

        .exposure(0.001f)
        .build(engine)

    view.colorGrading = colorGrading
    view.depthOfFieldOptions = dofOptions


    /// eventually render everything
 
 ```

**Comment by sortofsleepy** [comment]
Gonna mark Romain's answer as the correct response - thanks to the hint about looking at gamma correction I found some GLSL that is able to help get the textures back to looking much closer to the original. 

In case it happens to help anyone else, for reference that bit of code is is 


```glsl 
 vec3 gammaCorrection(vec3 diffuse, float gamma) {
        return pow(diffuse, vec3(1.0/gamma));
  }


```

**Reply by romainguy** [reply]
@sortofsleepy There's a much better way, easier, more efficient, and more correct. When you create your texture on the Filament side, you can tell Filament it's an sRGB texture and gamma correction will automatically happen when you sample the texture.

**Reply by romainguy** [reply]
More specifically use Texture.InternalFormat.SRGB8_A8

**Reply by romainguy** [reply]
You can also use the loadTexture() function from filament-android-utils. It'll do it for you.

**Reply by sortofsleepy** [reply]
@romainguy 

Ah! Good to know, thanks for following up. 

The shader based solution happened to work out with the structure of the project. I'm more or less a freelancer on it and the scene and related textures are partially constructed with an internal framework I don't have much say in changing; I'll make sure to make a note of it.

Thanks again!
